### PR TITLE
Add a `MultiFieldRetriever` to search content and metadata at the sametime.

### PR DESCRIPTION
**Pull request by DayuanJiang**
- **Description:** I created the `MultiFieldRetriever` to search not only the content of a document but also its metadata. The difference between it and the `Self-query Retriever` is that the `MultiFieldRetriever` merges the document's content and metadata and performs a single search, rather than searching the content first and then filtering the results by metadata. **You can see how it works in the [document notebook ](https://github.com/langchain-ai/langchain/blob/008618d38a3f8329c5aef4f2eb1f5da4f1e89401/docs/docs/modules/data_connection/retrievers/multi_field.ipynb)
  
  - **Dependencies:** No
  - **Tag maintainer:** @baskaryan 
  - **Twitter handle:** DayuanJian21687

### CODE REVIEW

**Comment by DayuanJiang** [comment]
@baskaryan 
If you think this PR is satisfactory, could you assist me with the linting issue? The error message is as follows:
```python
langchain/retrievers/multi_field.py:102: error: Item "BaseRetriever" of "Union[BaseRetriever, VectorStore]" has no attribute "from_documents"  [union-attr]
tests/unit_tests/retrievers/test_multi_field.py:25: error: Argument "retriever" to "from_documents" of "MultiFieldRetriever" has incompatible type "Type[TFIDFRetriever]"; expected 
```

In code 
```python
    @classmethod
    def from_documents(
        cls,
        retriever: Union[BaseRetriever, VectorStore],
        documents: List[Document],
        meta_data_keys: List[str] = [],
        search_kwargs: dict = {},
        str_concat_func: Callable = _create_content_and_metadata,
        **kwargs: Any,
    ) -> BaseRetriever:
 
        Returns:
            BaseRetriever: An instance of MultiFieldRetriever with the underlying
            retriever initialized with the processed documents.
        """
        pseudo_doc_list = [
            Document(
                page_content=str_concat_func(doc, meta_data_keys),
                metadata={"doc": doc},
            )
            for doc in documents
        ]

        retriever = retriever.from_documents(documents=pseudo_doc_list, **kwargs)
        if isinstance(retriever, VectorStore):
            return cls(retriever=retriever.as_retriever(search_kwargs=search_kwargs))
        return cls(retriever=retriever)
```

For the argument `retriever`, I intended to allow any subclass of `VectorStore`, such as `FAISS`, or any subclass of `Retriever`, such as `BM25Retriever`. However, it seems that the type hint is incorrect.

**Comment by DayuanJiang** [comment]
Hi @baskaryan , I hope you're doing well. When you have a chance, could you please take a look at this PR? Thank you for your time!

**Comment by DayuanJiang** [comment]
Hi @baskaryan, I hope all is well with you. When you get a moment, could you please review this pull request? I appreciate your time.

**Comment by DayuanJiang** [comment]
Hi @hwchase17, I hope all is well with you. When you get a moment, could you please review this pull request? I appreciate your time.

File: libs/langchain/langchain/retrievers/multi_field.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        """
+        pseudo_doc_list = [
+            Document(
+                page_content=str_concat_func(doc, meta_data_keys),
+                metadata={"doc": doc},
+            )
+            for doc in documents
+        ]
+
+        retriever = retriever.from_documents(documents=pseudo_doc_list, **kwargs)
```
Comment (hwchase17) [inline libs/langchain/langchain/retrievers/multi_field.py:102]: i dont think this typing works out. I think you want to probably assume that we are working with vectorstores

File: libs/langchain/langchain/retrievers/multi_field.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                `from_documents` method of the retriever or vector store.
+
+        Returns:
+            BaseRetriever: An instance of MultiFieldRetriever with the underlying
+            retriever initialized with the processed documents.
+        """
+        pseudo_doc_list = [
+            Document(
+                page_content=str_concat_func(doc, meta_data_keys),
+                metadata={"doc": doc},
```
Comment (hwchase17) [inline libs/langchain/langchain/retrievers/multi_field.py:97]: the whole document is a metadata field? that seems excessive...

**Comment by DayuanJiang** [comment]
@hwchase17  Thank you for your comment! I have resolved the linting issues and changed the document storage mechanism to use a dictionary, which seems more efficient compared to assigning all documents to the metadata.

**Comment by hwchase17** [comment]
i think this is best handled by data transformation ahead of time, and then ingestion into an existing index

**Pull request closed without merging**